File: 1 kyu/Faberge easter eggs crush test [linear]/solution.java
import java.math.BigInteger;

public class Faberge {

  private static BigInteger mo = BigInteger.valueOf(998244353);

  static final int MAX_NUMBER = 100_000;
  static final BigInteger[] MOD_INVERSES = new BigInteger[MAX_NUMBER + 1];

  // precompute modular inverses
  static {
    MOD_INVERSES[0] = BigInteger.ONE;
    for (int i = 1; i <= MAX_NUMBER; ++i) {
      MOD_INVERSES[i] = BigInteger.valueOf(i).modInverse(mo);
    }
  }

  public static BigInteger height(BigInteger n, BigInteger m) {

    // reduce number of tries according to modulus
    m = m.mod(mo);

    if (n.compareTo(m) >= 0) {
      return BigInteger.TWO.modPow(m, mo).subtract(BigInteger.ONE);
    }

    // sums up at most half of the binomial coefficients
    BigInteger flipped = null;
    if (BigInteger.TWO.multiply(n).compareTo(m) > 0) {
      n = m.subtract(n).subtract(BigInteger.ONE);
      flipped = BigInteger.TWO.modPow(m, mo).subtract(BigInteger.TWO);
    }


    BigInteger res = BigInteger.ZERO;
    BigInteger tmp = BigInteger.ONE;

    for (int i = 1; i <= n.intValue(); ++i) {
      tmp = tmp.multiply(m.add(BigInteger.valueOf(1 - i))).multiply(MOD_INVERSES[i]).mod(mo);
      res = res.add(tmp).mod(mo);
    }

    if (flipped != null) {
      return flipped.subtract(res).mod(mo);
    }
    return res;
  }
}
__________________________________________________
import java.math.BigInteger;

public class Faberge {
  private static BigInteger mo = BigInteger.valueOf(998244353);
  
  public static BigInteger height(BigInteger n, BigInteger m) {
    BigInteger ret = BigInteger.ZERO;
    BigInteger binom = BigInteger.ONE;
    for (BigInteger i = BigInteger.ONE; i.compareTo(n) <= 0; i = i.add(BigInteger.ONE)) {
      binom = binom.multiply(m.subtract(i).add(BigInteger.ONE)).multiply(i.modInverse(mo)).mod(mo);
      ret = ret.add(binom).mod(mo);
    }
    return ret;
  }

}
__________________________________________________
import java.math.BigInteger;

public class Faberge{
    private static BigInteger moduler = BigInteger.valueOf(998244353);

    public static BigInteger height(BigInteger n, BigInteger m){
        BigInteger sum = BigInteger.ZERO;
        BigInteger res = BigInteger.ONE;
        m = m.mod(moduler);

        for (BigInteger i = BigInteger.ONE; i.compareTo(n) <= 0; i = i.add(BigInteger.ONE)){
            res = (res.multiply(m.subtract(i).add(BigInteger.ONE))).mod(moduler)
                    .multiply(i.modPow(moduler.subtract(BigInteger.valueOf(2)), moduler));

            sum = sum.add(res).mod(moduler);
        }

        return sum;
    }
}

File: 1 kyu/Faberge easter eggs crush test [linear]/solution.kt
import java.math.BigInteger

object Faberge {
    val mod = 998244353.toBigInteger()

    fun height(n: BigInteger, m: BigInteger): BigInteger {
        if (n >= m) return (BigInteger.TWO.modPow(m, mod) - BigInteger.ONE) % mod
        var c = BigInteger.ONE
        var s = BigInteger.ZERO
        var i = BigInteger.ZERO
        while (i < n) {
            c = c * (m - i) * (i + BigInteger.ONE).modInverse(mod) % mod
            s += c
            ++i
        }
        return s % mod
    }
}
__________________________________________________
import java.math.BigDecimal
import java.math.BigInteger
import java.math.RoundingMode

object Faberge {
val mod = 998244353.toBigInteger()
    val preAllocate = mutableListOf(BigInteger.ZERO, BigInteger.ONE)

    init {
        for (haha_i in 2L until (80000L + 1L)) {
            preAllocate.add(
                mod - mod.toBigDecimal().divide(BigDecimal.valueOf(haha_i), RoundingMode.FLOOR)
                    .toBigInteger() * preAllocate[(mod % BigInteger.valueOf(haha_i)).toInt()] % mod
            )
        }
    }


    fun height(n: BigInteger, m: BigInteger): BigInteger {
        var h = BigInteger.ZERO
        var t = BigInteger.ONE
        val m1 = m % mod

        for (i in 1 until (n.toInt() + 1)) {
            t = t * (m1 - BigInteger.valueOf(i.toLong()) + BigInteger.ONE) * preAllocate[i] % mod
            h = (h + t) % mod
        }
        return h % mod
    }
}
__________________________________________________
import java.math.BigInteger

object Faberge {
    val MOD = 998244353.toBigInteger()

    fun height(n: BigInteger, m: BigInteger): BigInteger {
        var h = BigInteger.ZERO;
        var a = BigInteger.ONE;
        for (i in 1..n.intValueExact())
        {
            var inv = (i).toBigInteger().modInverse(MOD);
            a = a.multiply(m.add((1 - i).toBigInteger())).multiply(inv).mod(MOD)
            h = h.add(a).mod(MOD);
        }
        return h;
    }
}
__________________________________________________
import java.math.BigInteger
import java.math.BigInteger.*

object Faberge {
    val mo = 998244353.toBigInteger()

    fun height(n: BigInteger, m: BigInteger): BigInteger {
        var h = ZERO
        var t = ONE
        var i = ONE
        while (i <= n) {
            t = t * (m - i + ONE) * i.modInverse(mo) % mo
            h += t
            i += ONE
        }
        return h % mo
    }
}
__________________________________________________
import java.math.BigInteger
import java.math.BigInteger.ONE
import java.math.BigInteger.valueOf

object Faberge {
    private val mod = 998244353.toBigInteger()
    private val hahaInv: List<BigInteger> by lazy {
        val value = mutableListOf(0.toBigInteger(), 1.toBigInteger())
        for (i in (2..80000).map { it.toBigInteger() }) {
            value.add((mod.minus(mod.divide(i)).multiply(value[mod.mod(i).toInt()]).mod(mod)))
        }
        value
    }
    fun height(n: BigInteger, m: BigInteger): BigInteger {
        val m = m.mod(mod)
        var h = valueOf(0)
        var t = valueOf(1)
        for (i in (1..n.toLong()).map { valueOf(it) }) {
            t = t.multiply(m.minus(i).plus(ONE)).multiply(hahaInv[i.toInt()]).mod(mod);
            h = h.plus(t).mod(mod);
        }
        return h;
    }
}
